jax/lax/lax_parallel.py

```diff
@@ -579,6 +618,41 @@ def all_gather(x, axis_name):
   return _allgather(x, 0, psum(1, axis_name), axis_name)
 
 
+def _axis_index_translation_rule(c, *, axis_name, axis_env, platform):
+  div = xb.constant(c, np.array(axis_env.nreps // prod(axis_env.sizes),
+                                dtype=np.uint32))
+  mod = xb.constant(c, np.array(axis_env.sizes[-1], dtype=np.uint32))
+  unsigned_index = xops.Rem(xops.Div(xops.ReplicaId(c), div), mod)
+  return xops.ConvertElementType(unsigned_index, xb.dtype_to_etype(np.int32))
+
+def _axis_index_bind(*, axis_name):
+  dynamic_axis_env = pxla._thread_local_state.dynamic_axis_env
```

Comment:
Could add a comment for our future selves explaining that axis_index has a custom bind rule to make it work with vmap.